### ISSUE TITLE
Use a diffrent type of session cookie?

### ISSUE THREAD

**Issue by spec1re** [mailwatch/MailWatch#730]
I have a little annoying problem here, I run a few different services on the same box where MailWatch runs, so every time I log out of MailWatch I also get log out of my other Web interface and vice versa There must be some kind of php session conflict. 

Perhaps this could be a source for many other problems, which relay on php session/tokens and so on?
```

FreeBSD 10.3-RELEASE-p17
MailWatch Version 1.2.0
MailScanner Version 4.85.2
ClamAV Version 0.99.2
SpamAssassin Version 3.4.1
PHP Version 5.6.30
MySQL Version 10.1.21-MariaDB
NGINX 1.10.2
```

**Comment by asuweb** [comment]
We should consider using `session_name` to alleviate this issue.  

I suggest we set `session_name('MailWatch');` this will avoid these conflicts for cases list @spec1re.

@spec1re - I'm assuming these multiple services all run on the same domain?

**Comment by asuweb** [comment]
`session_name` can induce a performance hit, so we could make this optional via `conf.php`

**Comment by spec1re** [comment]
@asuweb - Yes, not same domain but same IP. eg. https://192.168.1.1:8080/

**Comment by asuweb** [comment]
@spec1re - can you try #734 please?  See comments in `conf.php.example`

**Comment by spec1re** [comment]
@asuweb - It looks like I need to download and install dev-version somehow?

**Comment by asuweb** [comment]
@spec1re - you can download a zip of this PR from here: https://github.com/asuweb/Mailwatch-1.2.0-asuantispam/archive/iss-730-session-conflict.zip

It's the latest develop version, with this PR

**Comment by spec1re** [comment]
@asuweb - Works like a champ and no performance drop, all is fast as usual. :)

**Comment by asuweb** [comment]
@spec1re - great, thanks!

**Comment by spec1re** [comment]
Just one more thing, where can I set the session timeout? At the moment, it looks like I get log out after 10 minutes or so.

**Comment by asuweb** [comment]
The session will stay alive if you are on the status page, as it auto refreshes.  Static pages will expire after a set time, this was implemented to enhance the security of the product.

**Comment by spec1re** [comment]
Okay, good to know thanks. :)

**Comment by spec1re** [comment]
Negative, If have the status page open and being on a different tab, I get log out after approx 10 minutes.

**Comment by asuweb** [comment]
@spec1re - that's strange.  Mine's been open for 24hrs without logout

**Comment by spec1re** [comment]
This is Firefox 52.0.2, maybe here is some kind of tab parking or so?

**Comment by stefaweb** [comment]
Works here with Chrome (Windows and Mac) and Safari Mac.

**Comment by spec1re** [comment]
Hm, so it look like I also get log out from my other web interface as well, so maybe its not just the MailWatch session. But I know for sure the other web interface do not have a session time out.

**Comment by asuweb** [comment]
Is it possible for you to try a different browser to see if it's browser specific or a more general issue?

**Comment by Skywalker-11** [comment]
Working here on Win10 with Firefox 51.0.2 and 52.0.2 (each 64-bit) when status.php is opened in another tab and even when window is minimized and other program is active.

**Comment by spec1re** [comment]
The problem is somewhere with SESSION_TIMEOUT, as soon I get log out I'm redirected to `login.php?error=timeout`
```

// Session garbage collection 5 minutes based on user activity
// or STATUS_REFRESH + 60 sec, whichever is greater.

if (defined(STATUS_REFRESH) && STATUS_REFRESH + 60 > 300) {
    ini_set('session.gc_maxlifetime', STATUS_REFRESH + 60);
    define('SESSION_TIMEOUT', STATUS_REFRESH + 59);
} else {
    ini_set('session.gc_maxlifetime', 300);
    define('SESSION_TIMEOUT', 299);
}
ini_set('session.gc_divisor', 1);
ini_set('session.gc_probability', 1);

$session_cookie_secure = false;
if (SSL_ONLY === true) {
    ini_set('session.cookie_secure', 1);
    $session_cookie_secure = true;
}
```
It looks like + 60 is not fail safe?

**Comment by asuweb** [comment]
Works in Chrome 56.0.2924.87, when minimised, different tabs etc

**Comment by stefaweb** [comment]
Tested here with Windows 10 and Firefox 51.0.1. Works.

**Comment by stefaweb** [comment]
Works with Firefox Mac 52.0.1.

**Comment by Skywalker-11** [comment]
The redirect to the login.php will happen a bit before the session would expire and earliest 5 min after opening/refreshing a page

@spec1re have you set STATUS_REFRESH  in your conf?

**Comment by spec1re** [comment]
@Skywalker-11 - Yes the default

```
// Default refresh rate in seconds for the Recent Messages screen
define('STATUS_REFRESH', 30);
```
I gave all +300 and no timeout anymore. :P

**Comment by spec1re** [comment]
Maybe a better approach is:

`ini_set('session.cookie_path', '/mailscanner/');`

the default is "/" which would include all the other web interfaces as well?

http://php.net/manual/en/session.configuration.php#ini.session.cookie-path

```
// more secure session cookies
ini_set('session.cookie_path', '/mailscanner/');
ini_set('session.use_cookies', 1);
ini_set('session.cookie_httponly', 0);
ini_set('session.use_only_cookies', 1);
ini_set('session.use_trans_sid', 0);
```

**Comment by spec1re** [comment]
Still get a timeout, most of the time while the page is reloading it self ...

**Comment by Skywalker-11** [comment]
Can you try the following:

1. open the status.php 
2. press F12 to open the debug console
3.  go to the settings there (yellow circle)
4. make sure the box for keeping the logs is enabled (red circle)
5. go to the network tab
6. go to another browser tab

If you go back after a few minutes there should be multiple calls of status.php in the network tab (second pic). If thats not the case firefox somehow stopps reloading the page.

![image](https://cloud.githubusercontent.com/assets/11180862/24791382/293a5104-1b7a-11e7-8701-82d2bf187544.png)

![image](https://cloud.githubusercontent.com/assets/11180862/24791649/3c922730-1b7b-11e7-8899-08cd6509f992.png)

**Comment by spec1re** [comment]
Okay, I'll try. Thanks again.

**Comment by Skywalker-11** [comment]
@spec1re did you solved the issue?

**Comment by spec1re** [comment]
Not really, I give all a extra 0 and now its fine for me:
```

// Session garbage collection 5 minutes based on user activity
// or STATUS_REFRESH + 60 sec, whichever is greater.

if (defined(STATUS_REFRESH) && STATUS_REFRESH + 600 > 3000) {
    ini_set('session.gc_maxlifetime', STATUS_REFRESH + 600);
    define('SESSION_TIMEOUT', STATUS_REFRESH + 590);
} else {
    ini_set('session.gc_maxlifetime', 3000);
    define('SESSION_TIMEOUT', 2990);
}
```
But I still think 59/60 = 1 sec. gap is maybe not enough.

**Comment by shawniverson** [comment]
Chiming in.  I think we can dump the garbage collection in favor of the SESSION_TIMEOUT approach instead and perhaps destroy the session cookie on timeout instead.  Although effective, the problem is that garbage collection is global.

**Comment by Skywalker-11** [comment]
How about seting a timestamp in the session and use that to manually check for expire. Would that solve the problem and still be secure?
eg something like this:
```
if (!isset($_SESSION['last_update']) || $_SESSION['last_update'] +SESSION_TIMEOUT < time()) {
  //timeout
} else {
  //session still valid
  $_SESSION['last_update'] = time(); 
}
```

**Comment by shawniverson** [comment]
I think that would be a good alternative.

**Comment by Skywalker-11** [comment]
@shawniverson do you want to make a pr? You are more into that session handling

**Comment by shawniverson** [comment]
Yes, I can do that, it will be a few hours before I can do this, if that is
okay.

**Comment by shawniverson** [comment]
Working on this now

**Comment by spec1re** [comment]
If there is something to test, just hit me up.

**Comment by shawniverson** [comment]
@spec1re Please test PR for me :)

**Comment by spec1re** [comment]
> Your session timed out. 

Cant login anymore. :P

**Comment by shawniverson** [comment]
lol hehe :P let me double check...

**Comment by shawniverson** [comment]
Huh, mines working....whats your STATUS_REFRESH in conf.php?

**Comment by spec1re** [comment]
30, let me check the patch again, maybe I've overlooked something.

**Comment by spec1re** [comment]
`Notice: A session had already been started - ignoring session_start() in /usr/local/www/mailscanner/status.php on line 36 Call Stack: 0.0001 228104 1. {main}() /usr/local/www/mailscanner/status.php:0 0.0184 1736392 2. session_start() /usr/local/www/mailscanner/status.php:36 Warning: Cannot modify header information - headers already sent by (output started at /usr/local/www/mailscanner/status.php:36) in /usr/local/www/mailscanner/login.function.php on line 43 Call Stack: 0.0001 228104 1. {main}() /usr/local/www/mailscanner/status.php:0 0.0187 1737208 2. require('/usr/local/www/mailscanner/login.function.php') /usr/local/www/mailscanner/status.php:37 0.0187 1737392 3. header() /usr/local/www/mailscanner/login.function.php:43 `

getting closer..

**Comment by shawniverson** [comment]
ahh, I see your issue, you are not on latest develop.  session_start() was recently moved into functions.php

**Comment by shawniverson** [comment]
which just so happens to be necessary for this to work properly

**Comment by spec1re** [comment]
BTW Im running Mailwatch v1.2.3-dev, but you patch is for 1.2.0 ?

**Comment by shawniverson** [comment]
Nah, it is the latest, the 1.2.0 is because I haven't renamed my fork.

**Comment by spec1re** [comment]
So I need to update to the latest https://github.com/mailwatch/MailWatch/tree/develop first?

If so, I'll try it later this afternoon.

**Comment by shawniverson** [comment]
Yes.  I put a hold on merging while we test this further, so no rush.

**Comment by spec1re** [comment]
Just one thing, is there any nogo to run the latest develop build, because this is my productive setup. :D

**Comment by shawniverson** [comment]
I am running it in production right now at commit 409e3fc2c5e1c9629a319a2a345e0ba15b82b4af.  I won't swear to any future commits,, if any happen to merge.  The only issue I've witnessed so far (besides this issue), is Issue #742, which may or may not impact you.  It appears that it regressed sometime after 1.2.0-RC4.   If you are already on a 1.2.3-dev prior to this commit, changes so far have been fairly minimal.

**Comment by spec1re** [comment]
Okay, thanks. :)

**Comment by endelwar** [comment]
If you want to be sure to run the exact commit Shawn is working on use these commands:
```bash
git checkout -b shawniverson-sessiontimeoutfix develop
git pull git://github.com/shawniverson/1.2.0.git sessiontimeoutfix
```

**Comment by spec1re** [comment]
Okay, running latest dev + sessiontimeoutfix (incl. PR from few minutes a go) :P

**Comment by spec1re** [comment]
Still get a session timeout and redirect to `login.php?error=timeout`

**Comment by shawniverson** [comment]
On the Recent Messages pane?

**Comment by spec1re** [comment]
Yes, status.php

**Comment by shawniverson** [comment]
And it is refreshing every 30 seconds?

**Comment by spec1re** [comment]
Exactly, in one tab I have Recent Messages in the other one pfSense web interface with the same URL/IP open.

**Comment by shawniverson** [comment]
Is this running on pfsense?  If so which version?  (BTW I find this is an odd combination)

**Comment by spec1re** [comment]
```
pfSense 2.3.3-RELEASE-p1
FreeBSD 10.3-RELEASE-p17
MailWatch Version 1.2.3-dev
MailScanner Version 4.85.2
ClamAV Version 0.99.2
SpamAssassin Version 3.4.1
PHP Version 5.6.30
MySQL Version 10.1.21-MariaDB
NGINX 1.10.2
```

**Comment by shawniverson** [comment]
ok, well, good news, mine just timed out too 😃

**Comment by spec1re** [comment]
With these values, it never times out:

```
// Session timeout 5 minutes based on user activity
// or STATUS_REFRESH + 60 sec, whichever is greater.

if (defined(STATUS_REFRESH) && STATUS_REFRESH + 600 > 3000) {
    define('SESSION_TIMEOUT', STATUS_REFRESH + 600);
} else {
    define('SESSION_TIMEOUT', 3000);
}
```
👶 🍭

**Comment by shawniverson** [comment]
Yeah, but that defeats the purpose 😄 with those values, there's little security advantage.  Need to ideally work within the 5 min restraint...

**Comment by shawniverson** [comment]
Interesting, just out of curiousity, which of your windows is the "active" window while testing this?  I'm trying to figure out what I did to make it time out.  Grrrrr........

**Comment by spec1re** [comment]
To be honest I don't know, it feels totally random for me.

![grafik](https://cloud.githubusercontent.com/assets/11883076/24912178/e8ccf59a-1ebc-11e7-8d02-3b915ae4e3ee.png)

**Comment by shawniverson** [comment]
Yeah, I'm going to assume it is a timing issue, and make an adjustment to the amount headroom and have you try it.

**Comment by shawniverson** [comment]
Ok I adjusted the PR (doubled the timer to 10 min)....see if that makes a difference.

**Comment by spec1re** [comment]
Same here, I've set 10 min, no timeout for now...

**Comment by shawniverson** [comment]
Yeah, going to run it for the rest of the day and see what happens.....if we have to go 10 min, then we may just have to accept that as a reasonable compromise.

**Comment by spec1re** [comment]
10 min looks like a good value, no hiccups for me. :)

**Comment by stefaweb** [comment]
No problem here with the new version.
I prefer 10  minutes. ;-)

**Comment by Skywalker-11** [comment]
Can we close this as it is applied by #744 ?

**Comment by Daniel-Beardsmore** [comment]
Just to be clear on this: the issue with v1.2.3-dev that I see—which is not mentioned here explicitly—is that leaving Recent Messages and any static page (e.g. Tools and Links) open in separate tabs causes the static page to log out the session regardless of the self-refresh functionality of Recent Messages. This is now resolved, I'm guessing? (This is v1.2.3 dev in EFA 3.0.2.2, where the session cookie is still called PHPSESSID.)

Also, the "Your session has timed out." page can't be used (it reports "Error: unable to validate security token" when trying to log back in) and none of the login pages carry the URL that you were on before to restore you to the page that you were on. The token error occurs in both Firefox and Opera.

I didn't want to open a new issue if this is now resolved.

**Comment by shawniverson** [comment]
Having multiple tabs open in different sections of MailWatch will cause
both session issues and token issues.  The reason is that there is only one
session, not a separate session for each tab.  If working on multiple
sections of MailWatch concurrently , using a Private or Incognito window
will allow two sessions to exist concurrently.  This probably needs
documented.



On Wed, May 10, 2017 at 8:47 AM, Daniel Beardsmore <notifications@github.com
